### PR TITLE
Use absolute page paths from page mappings for the middleware SSR loader

### LINKED ISSUE #30714: When concurrentFeatures is enabled, context provider in _app file does not seem to work properly
### What version of Next.js are you using?

^12.0.2-canary.14

### What version of Node.js are you using?

v16.13.0

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

tested on localhost

### Describe the Bug

* I enabled experimental concurrentFeatures in my `next.config.js`
* I created custom app file (_app.js) and it includes `QueryClientProvider` provided `react-query`. And I expected It will allow the application to use queryClient in the Page Component.
* However, I am getting the error like below.
```ReferenceError: useQueryClient is not defined
    at Home (webpack://_N_E/./pages/index.js?:18:25)
    at renderWithHooks (webpack://_N_E/./node_modules/react-dom/cjs/react-dom-server.browser.development.js?:5212:16)
    at renderIndeterminateComponent (webpack://_N_E/./node_modules/react-dom/cjs/react-dom-server.browser.development.js?:5285:15)
    at renderElement (webpack://_N_E/./node_modules/react-dom/cjs/react-dom-server.browser.development.js?:5467:7)
    at renderNodeDestructive (webpack://_N_E/./node_modules/react-dom/cjs/react-dom-server.browser.development.js?:5606:11)
    at renderContextProvider (webpack://_N_E/./node_modules/react-dom/cjs/react-dom-server.browser.development.js?:5441:3)
    at renderElement (webpack://_N_E/./node_modules/react-dom/cjs/react-dom-server.browser.development.js?:5538:11)
    at renderNodeDestructive (webpack://_N_E/./node_modules/react-dom/cjs/react-dom-server.browser.development.js?:5606:11)
    at renderContextProvider (webpack://_N_E/./node_modules/react-dom/cjs/react-dom-server.browser.development.js?:5441:3)
    at renderElement (webpack://_N_E/./node_modules/react-dom/cjs/react-dom-server.browser.development.js?:5538:11)
```

### Expected Behavior

I expected it would work properly. If I have misunderstanding on the feature. Please let me know.

### To Reproduce

https://github.com/seonghyeonkimm/issue-nextjs-provider

* clone source code from [this repository](https://github.com/seonghyeonkimm/issue-nextjs-provider)
* install dependencies and try to run `yarn dev`
* next app throw the error mentioned above.

### PR DESCRIPTION
**Pull request by shuding**
This also removes the fallback Document, so it's required to have a function Document now (but this is mentioned in the docs already).

Fixes #30714.

## Bug

- [x] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Errors have helpful link attached, see `contributing.md`

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [ ] Make sure the linting passes by running `yarn lint`

### CODE REVIEW

File: packages/next/build/webpack/loaders/next-middleware-ssr-loader/index.ts [hunk 1]
```diff
@@ -1,34 +1,11 @@
...
-  }
-  return has
-}
-
 export default async function middlewareRSCLoader(this: any) {
   const {
+    absoluteAppPath,
+    absoluteDocumentPath,
+    // absoluteErrorPath,
+    // absolute404Path,
```
Comment (huozhi) [inline packages/next/build/webpack/loaders/next-middleware-ssr-loader/index.ts:8]:
```suggestion
```

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | shuding/next.js shu/39b3 | Change |
| - | - | - | - |
| buildDuration | 26.2s | 26.5s | ⚠️ +278ms |
| buildDurationCached | 5.2s | 5.2s | -44ms |
| nodeModulesSize | 293 MB | 293 MB | -576 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | shuding/next.js shu/39b3 | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 4.345 | 4.337 | -0.01 |
| / avg req/sec | 575.38 | 576.39 | +1.01 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 2.283 | 2.297 | ⚠️ +0.01 |
| /error-in-render avg req/sec | 1095.22 | 1088.41 | ⚠️ -6.81 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | shuding/next.js shu/39b3 | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 28 kB | 28 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 71.9 kB | 71.9 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | shuding/next.js shu/39b3 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | shuding/next.js shu/39b3 | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.23 kB | 1.23 kB | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 2.38 kB | 2.38 kB | ✓ |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 635 B | 635 B | ✓ |
| image-HASH.js gzip | 4.44 kB | 4.44 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 1.87 kB | 1.87 kB | ✓ |
| routerDirect..HASH.js gzip | 321 B | 321 B | ✓ |
| script-HASH.js gzip | 383 B | 383 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 334f979574ae..6f4.css gzip | 106 B | 106 B | ✓ |
| Overall change | 13.1 kB | 13.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | shuding/next.js shu/39b3 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | shuding/next.js shu/39b3 | Change |
| - | - | - | - |
| index.html gzip | 522 B | 522 B | ✓ |
| link.html gzip | 533 B | 533 B | ✓ |
| withRouter.html gzip | 516 B | 516 B | ✓ |
| Overall change | 1.57 kB | 1.57 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | shuding/next.js shu/39b3 | Change |
| - | - | - | - |
| buildDuration | 28.4s | 28s | -312ms |
| buildDurationCached | 5.1s | 5.2s | ⚠️ +29ms |
| nodeModulesSize | 293 MB | 293 MB | -576 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | shuding/next.js shu/39b3 | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 4.398 | 4.38 | -0.02 |
| / avg req/sec | 568.49 | 570.76 | +2.27 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 2.319 | 2.438 | ⚠️ +0.12 |
| /error-in-render avg req/sec | 1078.28 | 1025.41 | ⚠️ -52.87 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | shuding/next.js shu/39b3 | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.2 kB | 28.2 kB | ✓ |
| webpack-HASH.js gzip | 1.43 kB | 1.43 kB | ✓ |
| Overall change | 72.1 kB | 72.1 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | shuding/next.js shu/39b3 | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | shuding/next.js shu/39b3 | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.22 kB | 1.22 kB | ✓ |
| _error-HASH.js gzip | 180 B | 180 B | ✓ |
| amp-HASH.js gzip | 305 B | 305 B | ✓ |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.38 kB | 2.38 kB | ✓ |
| head-HASH.js gzip | 342 B | 342 B | ✓ |
| hooks-HASH.js gzip | 622 B | 622 B | ✓ |
| image-HASH.js gzip | 4.46 kB | 4.46 kB | ✓ |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 1.91 kB | 1.91 kB | ✓ |
| routerDirect..HASH.js gzip | 314 B | 314 B | ✓ |
| script-HASH.js gzip | 375 B | 375 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 334f979574ae..6f4.css gzip | 106 B | 106 B | ✓ |
| Overall change | 13.1 kB | 13.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | shuding/next.js shu/39b3 | Change |
| - | - | - | - |
| _buildManifest.js gzip | 460 B | 460 B | ✓ |
| Overall change | 460 B | 460 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | shuding/next.js shu/39b3 | Change |
| - | - | - | - |
| index.html gzip | 522 B | 522 B | ✓ |
| link.html gzip | 535 B | 535 B | ✓ |
| withRouter.html gzip | 516 B | 516 B | ✓ |
| Overall change | 1.57 kB | 1.57 kB | ✓ |

</details>

</details>
Commit: d4943af5949ef20b41c4d69478cf76613a39314e

**Comment by ijjk** [comment]
## Failing test suites 
Commit: d4943af5949ef20b41c4d69478cf76613a39314e

**test/integration/react-streaming-and-server-components/test/index.test.js** 
- CSS dev > should include global styles under `concurrentFeatures: true`
- CSS dev > should include global styles with `serverComponents: true`
- CSS prod > should include global styles under `concurrentFeatures: true`
- CSS prod > should include global styles with `serverComponents: true`
- concurrentFeatures - dev > should render the correct html
- concurrentFeatures - dev > should suspense next/link on server side
- concurrentFeatures - dev > should suspense next/image on server side
- concurrentFeatures - dev > should support multi-level server component imports
- concurrentFeatures - prod > should render the correct html
- concurrentFeatures - prod > should suspense next/link on server side
- concurrentFeatures - prod > should suspense next/image on server side
- concurrentFeatures - prod > should support multi-level server component imports

<details>
<summary>Expand output</summary>

  ● concurrentFeatures - prod › should render the correct html

    expect(received).toContain(expected) // indexOf

    Expected substring: "thisistheindexpage.server"
    Received string:    "Error: `Document.getInitialProps` is not supported in the Edge Runtime, please make sure pages/_document exports a function component."

      230 |
      231 | function runSuite(suiteName, env, { runTests, before, after }) {
    > 232 |   const context = { appDir }
          |                             ^
      233 |   describe(`${suiteName} ${env}`, () => {
      234 |     if (env === 'prod') {
      235 |       beforeAll(async () => {

      at Object.<anonymous> (integration/react-streaming-and-server-components/test/index.test.js:232:29)

  ● concurrentFeatures - prod › should suspense next/link on server side

    expect(received).toContain(expected) // indexOf

    Expected substring: "go home"
    Received string:    ""

      239 |       })
      240 |     }
    > 241 |     if (env === 'dev') {
          |                      ^
      242 |       beforeAll(async () => {
      243 |         before?.()
      244 |         context.appPort = await findPort()

      at Object.<anonymous> (integration/react-streaming-and-server-components/test/index.test.js:241:22)

  ● concurrentFeatures - prod › should suspense next/image on server side

    expect(received).toContain(expected) // indexOf

    Matcher error: received value must not be null nor undefined

    Received has value: undefined

      245 |         context.server = await nextDev(context.appDir, context.appPort)
      246 |       })
    > 247 |     }
          |      ^
      248 |     afterAll(async () => {
      249 |       after?.()
      250 |       await killApp(context.server)

      at Object.<anonymous> (integration/react-streaming-and-server-components/test/index.test.js:247:6)

  ● concurrentFeatures - prod › should support multi-level server component imports

    expect(received).toContain(expected) // indexOf

    Expected substring: "bar.server.js:"
    Received string:    "Error: `Document.getInitialProps` is not supported in the Edge Runtime, please make sure pages/_document exports a function component."

      249 |       after?.()
      250 |       await killApp(context.server)
    > 251 |     })
          |      ^
      252 |
      253 |     runTests(context)
      254 |   })

      at Object.<anonymous> (integration/react-streaming-and-server-components/test/index.test.js:251:6)

  ● concurrentFeatures - dev › should render the correct html

    expect(received).toContain(expected) // indexOf

    Expected substring: "thisistheindexpage.server"
    Received string:    "Error: `Document.getInitialProps` is not supported in the Edge Runtime, please make sure pages/_document exports a function component."

      230 |
      231 | function runSuite(suiteName, env, { runTests, before, after }) {
    > 232 |   const context = { appDir }
          |                             ^
      233 |   describe(`${suiteName} ${env}`, () => {
      234 |     if (env === 'prod') {
      235 |       beforeAll(async () => {

      at Object.<anonymous> (integration/react-streaming-and-server-components/test/index.test.js:232:29)

  ● concurrentFeatures - dev › should suspense next/link on server side

    expect(received).toContain(expected) // indexOf

    Expected substring: "go home"
    Received string:    ""

      239 |       })
      240 |     }
    > 241 |     if (env === 'dev') {
          |                      ^
      242 |       beforeAll(async () => {
      243 |         before?.()
      244 |         context.appPort = await findPort()

      at Object.<anonymous> (integration/react-streaming-and-server-components/test/index.test.js:241:22)

  ● concurrentFeatures - dev › should suspense next/image on server side

    expect(received).toContain(expected) // indexOf

    Matcher error: received value must not be null nor undefined

    Received has value: undefined

      245 |         context.server = await nextDev(context.appDir, context.appPort)
      246 |       })
    > 247 |     }
          |      ^
      248 |     afterAll(async () => {
      249 |       after?.()
      250 |       await killApp(context.server)

      at Object.<anonymous> (integration/react-streaming-and-server-components/test/index.test.js:247:6)

  ● concurrentFeatures - dev › should support multi-level server component imports

    expect(received).toContain(expected) // indexOf

    Expected substring: "bar.server.js:"
    Received string:    "Error: `Document.getInitialProps` is not supported in the Edge Runtime, please make sure pages/_document exports a function component."

      249 |       after?.()
      250 |       await killApp(context.server)
    > 251 |     })
          |      ^
      252 |
      253 |     runTests(context)
      254 |   })

      at Object.<anonymous> (integration/react-streaming-and-server-components/test/index.test.js:251:6)

  ● CSS dev › should include global styles under `concurrentFeatures: true`

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `CSS dev should include global styles under \`concurrentFeatures: true\` 1`

    Snapshot: "rgb(255, 0, 0)"
    Received: null

       8 |       `window.getComputedStyle(document.querySelector('#red')).color`
       9 |     )
    > 10 |     expect(currentColor).toMatchInlineSnapshot(`"rgb(255, 0, 0)"`)
         |                          ^
      11 |   })
      12 |   it('should include global styles with `serverComponents: true`', async () => {
      13 |     const browser = await webdriver(context.appPort, '/global-styles-rsc')

      at Object.<anonymous> (integration/react-streaming-and-server-components/test/css.js:10:26)

  ● CSS dev › should include global styles with `serverComponents: true`

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `CSS dev should include global styles with \`serverComponents: true\` 1`

    Snapshot: "rgb(255, 0, 0)"
    Received: null

      15 |       `window.getComputedStyle(document.querySelector('#red')).color`
      16 |     )
    > 17 |     expect(currentColor).toMatchInlineSnapshot(`"rgb(255, 0, 0)"`)
         |                          ^
      18 |   })
      19 |   // TODO: fix this test
      20 |   // it.skip('should include css modules with `serverComponents: true`', async () => {

      at Object.<anonymous> (integration/react-streaming-and-server-components/test/css.js:17:26)

  ● CSS prod › should include global styles under `concurrentFeatures: true`

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `CSS prod should include global styles under \`concurrentFeatures: true\` 1`

    Snapshot: "rgb(255, 0, 0)"
    Received: null

       8 |       `window.getComputedStyle(document.querySelector('#red')).color`
       9 |     )
    > 10 |     expect(currentColor).toMatchInlineSnapshot(`"rgb(255, 0, 0)"`)
         |                          ^
      11 |   })
      12 |   it('should include global styles with `serverComponents: true`', async () => {
      13 |     const browser = await webdriver(context.appPort, '/global-styles-rsc')

      at Object.<anonymous> (integration/react-streaming-and-server-components/test/css.js:10:26)

  ● CSS prod › should include global styles with `serverComponents: true`

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `CSS prod should include global styles with \`serverComponents: true\` 1`

    Snapshot: "rgb(255, 0, 0)"
    Received: null

      15 |       `window.getComputedStyle(document.querySelector('#red')).color`
      16 |     )
    > 17 |     expect(currentColor).toMatchInlineSnapshot(`"rgb(255, 0, 0)"`)
         |                          ^
      18 |   })
      19 |   // TODO: fix this test
      20 |   // it.skip('should include css modules with `serverComponents: true`', async () => {

      at Object.<anonymous> (integration/react-streaming-and-server-components/test/css.js:17:26)

</details>

**Review by huozhi — COMMENTED** [review]
note: try adding _document.server as fallback

**Pull request closed without merging**